File: haskell/algebraic-polynomials/src/Math/Internal/Positive.hs
-- |
-- Module      : Math.Internal.Positive
-- Description : Strictly positive integers with type-level safety
-- Stability   : experimental
--
-- A newtype wrapper for strictly positive integers (\(> 0\)), encoding the
-- positivity invariant in the type system. This eliminates runtime errors
-- for functions that require positive input (e.g., 'Math.Internal.PrimeFactors.factorise',
-- 'Math.Polynomial.Cyclotomic.cyclotomic').
--
-- Construct values via 'positive' (checked) or numeric literals (via the
-- 'Num' instance, which errors at runtime for non-positive values -- but
-- literals are known at write time).
module Math.Internal.Positive
  ( Positive
  , unPositive
  , positive
  , unsafePositive
  ) where

import Numeric.Natural (Natural)

-- | A strictly positive integer (\(> 0\)).
--
-- The internal representation is 'Natural', which is guaranteed to be
-- non-negative; the 'Positive' wrapper further guarantees it is nonzero.
--
-- Use 'positive' for safe construction, or 'unsafePositive' when
-- positivity is guaranteed by context.
newtype Positive = Positive { unPositive :: Natural }
  deriving (Eq, Ord, Show)

-- | Safe constructor: returns @'Just' p@ for positive input, @'Nothing'@
-- for zero.
--
-- >>> positive 5
-- Just (Positive {unPositive = 5})
-- >>> positive 0
-- Nothing
positive :: Natural -> Maybe Positive
positive 0 = Nothing
positive n = Just (Positive n)

-- | Unsafe constructor for cases where positivity is guaranteed by
-- context (e.g., the result of @a + 1@ where @a@ is a 'Natural').
--
-- Throws an error if given zero.
unsafePositive :: Natural -> Positive
unsafePositive 0 = error "unsafePositive: zero"
unsafePositive n = Positive n

-- | 'Num' instance for 'Positive'.
--
-- 'fromInteger' errors on non-positive values, making it safe for
-- numeric literals (which are known at write time).
--
-- Addition and multiplication preserve positivity.
-- 'negate' always errors since the result would be non-positive.
instance Num Positive where
  Positive a + Positive b = Positive (a + b)
  Positive a * Positive b = Positive (a * b)
  abs = id
  signum _ = Positive 1
  negate _ = error "Positive.negate: result would be non-positive"
  fromInteger n
    | n > 0     = Positive (fromInteger n)
    | otherwise = error $ "Positive.fromInteger: non-positive value " ++ show n

File: scala3/src/main/scala/surd/Positive.scala
package surd

/** A strictly positive integer (> 0).
  *
  * Implemented as an opaque type over BigInt for zero-cost abstraction.
  * Construct via `Positive.apply` (checked) or `Positive.unsafe` (unchecked).
  */
opaque type Positive = BigInt

object Positive:
  /** Smart constructor: returns None for non-positive values. */
  def apply(n: BigInt): Option[Positive] =
    if n > 0 then Some(n) else None

  /** Smart constructor from Int. */
  def apply(n: Int): Option[Positive] = apply(BigInt(n))

  /** Unsafe constructor for cases where positivity is guaranteed.
    * Throws IllegalArgumentException for non-positive values.
    */
  def unsafe(n: BigInt): Positive =
    require(n > 0, s"Positive.unsafe: non-positive value $n")
    n

  def unsafe(n: Int): Positive = unsafe(BigInt(n))

  /** Literal constructor for known-at-write-time constants. */
  def literal(n: Int): Positive =
    require(n > 0, s"Positive.literal: non-positive value $n")
    BigInt(n)

  extension (p: Positive)
    def value: BigInt = p
    def toInt: Int = p.toInt
    def toLong: Long = p.toLong

    def +(q: Positive): Positive = p + q
    def *(q: Positive): Positive = p * q
